src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -144,12 +165,33 @@ public Object invokeMethod(String name, Object args) {
                     throw new GroovyRuntimeException(x);
                 }
             }
+            Set<String> ambiguousFunctionNames = new HashSet<>(0);
             for (StepDescriptor d : StepDescriptor.all()) {
-                functions.put(d.getFunctionName(), d);
+                String functionName = d.getFunctionName();
+                if (functions.containsKey(functionName)) {
+                    ambiguousFunctionNames.add(functionName);
+                }
+                // TODO: Switch we switch to putIfAbsent so that the descriptor with the highest ordinal value is preferred for ambiguous functions?
+                functions.put(functionName, d);
+                stepClassNames.put(d.clazz.getName(), d);
             }
+            ambiguousFunctions = functions.keySet().stream()
+                    .filter(ambiguousFunctionNames::contains)
+                    .collect(Collectors.groupingBy(Function.identity(),
+                            Collectors.mapping(n -> functions.get(n).clazz.getName(), Collectors.toList())));
+            ambiguousFunctionsLogged = new HashSet<>(0);
         }
-        final StepDescriptor sd = functions.get(name);
+        final StepDescriptor sd = functions.getOrDefault(name, stepClassNames.get(name));
         if (sd != null) {
+            List<String> ambiguousClassNames = ambiguousFunctions.get(name);
+            if (ambiguousClassNames != null && ambiguousFunctionsLogged.add(name)) {
+                LOGGER.log(Level.WARNING, "Invoking ambiguous Pipeline Step ‘{0}’ ({1}) in {3}. " + 
```

Comment:
I would prefer to send this to the log of the the running build instead, but to do so, we need the `StepContext` in `invokeStep`, but `invokeStep` has no way to know if the step was invoked using the ambiguous function name or the unambiguous class name. We could change the signature of `invokeStep` to include that information, but this class is public, and the method is private, so I'm not sure if that would be ok. My other thought was to store the name used to invoke the step in a field on the class so we could use it in `invokeStep`, but that would be completely broken if `invokeMethod` can be called concurrently. Thoughts?

Reply:
Maybe we could do something like 
```
CpsThread thread = CpsThread.current();
if (thread != null) {
  TaskListener listener = thread.getContextVariable(TaskListener.class)
  if (listener != null) {
    listener.getLogger().println("MESSAGE HERE")
  }
}
```

Reply:
ha, I forgot semicolons.

Reply:
I'll try that, thanks!

Reply:
It looks like that doesn't quite work, because `thread.getContextVariable(TaskListener.class)` is null (and from looking through usages I think it could only be non-null inside of an async CpsBodyExecution). It also looks like [DefaultStepContext#get](https://github.com/jenkinsci/workflow-support-plugin/blob/8105510ba5617c19e90c8cf5ee467fb559833d8d/src/main/java/org/jenkinsci/plugins/workflow/support/DefaultStepContext.java#L73) is the place that constructs TaskListeners from scratch, so I think we do need an actual context.

Reply:
Well, boo. Then maybe change the signature for `invokeStep` but also add an overload with the existing signature, which would call the new signature with `null` for the step name?

Reply:
Ok, I'll take that approach for now, thanks!

Reply:
(done in 90e1e1b)